include/unifex/get_exception_ptr.hpp

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/exception.hpp>
+#include <unifex/tag_invoke.hpp>
+#include <unifex/type_traits.hpp>
+
+#include <memory>
+#include <system_error>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex
+{
+  namespace _get_exception_ptr
+  {
+    inline const struct _fn {
+      // forward std::exception_ptr
+      std::exception_ptr operator()(std::exception_ptr &&eptr) const noexcept {
+        return std::forward<std::exception_ptr>(eptr);
+      }
+
+      // convert std::exception based types to std::exception_ptr
+      template(typename Exception)(
+          requires std::is_base_of_v<std::exception, Exception>)
+          std::exception_ptr
+          operator()(Exception &&ex) const noexcept {
+        return make_exception_ptr(std::forward<Exception>(ex));
+      }
+
+      // use customization point
+      // to resolve ErrorCode -> std::exception_ptr conversion
+      template(typename ErrorCode)
+          (requires is_tag_invocable_v<_fn, ErrorCode>)
+      std::exception_ptr operator()(ErrorCode &&error) const noexcept {
+        return tag_invoke(*this, std::forward<ErrorCode>(error));
```

Comment:
I think we should assert here that the selected `tag_invoke` overload is `noexcept` (although that is a tough contract to meet since most exceptions don't have `noexcept` constructors). I see this `operator()` is `noexcept` so we're getting termination if any `tag_invoke` overload for `get_exception_ptr` exits with an exception. Which, meh, happens in synchronous code when you try to throw an exception with another one active. Dunno. I'm undecided whether to push this trouble off on our users or not.

Reply:
My opinion is that we should assume that an `ErrorCode` to ` exception_ptr` must be `noexcept`, thus, a static_assert might be helpful (or a new `nothow_tag_invokable` concept).